File: algorithm_library/python/data-structure/Mos_algorithm.py
def mos_algorithm(data, queries):
    """
    data : list of int : 元の配列
    queries : list of tuple : クエリのリスト
    クエリは 0-indexed で半開区間[l,r)
    """
    from math import sqrt, ceil
    n = len(data)
    q = len(queries)
    B = ceil(sqrt(3)*n / (sqrt(2)*sqrt(q))) # bukectのサイズ(nyaanさんの記事より)
    L,R = 0,0
    # queryの順番を格納する用
    order = [-1] * q

    ########################################
    # 状態を初期化する処理を書く

    ########################################

    def _add(i, ans):
        # i番目の要素を含めて考えるときへstatesを更新
        ########################################

        return ans
        ########################################

    def _delete(i, ans):
        # i番目の要素を除いて考えるときへstatesを更新
        ########################################

        return ans
        ########################################

    # クエリの解答順
    # (ceil(l/B), r*-1^{ceil(l/B)}) の昇順にするのが最速ぽい
    for idx, (l,r) in enumerate(queries):
        order[idx] = ((l//B)<<40)+((r if (l//B)&1 else -r)<<20)+idx
    order.sort()

    mask = (1<<20)-1
    ret = [-1] * q
    for lri in order:  # クエリに答えていく
        i = lri & mask
        l,r = queries[i]
        
        # クエリ[l,r)に対してstatesを更新する
        while R < r:  # rまで伸長
            ans = _add(R,ans)
            R += 1
        while R > r:  # rまで短縮
            R -= 1
            ans = _delete(R,ans)
            
        while L < l:  # lまで短縮
            ans = _delete(L,ans)
            L += 1
        while L > l:  # lまで伸長
            L -= 1
            ans = _add(L,ans)
        
        ########################################
        # クエリに答える作業をここで書く
        # ret[i] に i 番目のクエリの答えを格納する
        ret[i] = ans
        ########################################
        
    return ret

File: algorithm_library/cpp-from-py/data-structure/Mos_algorithm.cpp
#include <bits/stdc++.h>
using namespace std;

template <class Add, class Del, class Answer>
auto mos_algorithm(int n, const vector<pair<int, int>>& queries, Add add, Del del, Answer answer) {
    int q = (int)queries.size();
    int B = max(1, (int)ceil(sqrt(3.0) * n / (sqrt(2.0) * sqrt(max(1, q)))));
    vector<int> order(q);
    iota(order.begin(), order.end(), 0);
    sort(order.begin(), order.end(), [&](int i, int j) {
        auto [li, ri] = queries[i];
        auto [lj, rj] = queries[j];
        int bi = li / B, bj = lj / B;
        if (bi != bj) return bi < bj;
        return (bi & 1) ? ri < rj : ri > rj;
    });
    using R = decltype(answer(0));
    vector<R> ret(q);
    int L = 0, Rr = 0;
    for (int i : order) {
        auto [l, r] = queries[i];
        while (Rr < r) add(Rr++);
        while (Rr > r) del(--Rr);
        while (L < l) del(L++);
        while (L > l) add(--L);
        ret[i] = answer(i);
    }
    return ret;
}
